datanode/api/trading_data_v2.go

```diff
@@ -1115,13 +1159,65 @@ func (t *tradingDataServiceV2) ObservePositions(req *v2.ObservePositionsRequest,
 		t.log.Debug("Positions subscriber - new rpc stream", logging.Uint64("ref", ref))
 	}
 
-	return observe(ctx, t.log, "Position", positionsChan, ref, func(position entities.Position) error {
+	return observeBatch(ctx, t.log, "Position", positionsChan, ref, func(positions []entities.Position) error {
+		protos := make([]*vega.Position, len(positions))
+		for i := 0; i < len(positions); i++ {
+			protos[i] = positions[i].ToProto()
+		}
+
 		return srv.Send(&v2.ObservePositionsResponse{
-			Position: position.ToProto(),
+			Positions: protos,
 		})
 	})
 }
 
+func (t *tradingDataServiceV2) sendPositionsInitialImage(ctx context.Context, req *v2.ObservePositionsRequest, srv v2.TradingDataService_ObservePositionsServer) error {
+	var positions []entities.Position
+	var err error
+
+	// By market and party
+	if req.PartyId != nil && req.MarketId != nil {
+		position, err := t.positionService.GetByMarketAndParty(ctx, *req.MarketId, *req.PartyId)
+		if err != nil {
+			return errors.Wrap(err, "getting initial positions by market+party")
+		}
+		positions = append(positions, position)
+	}
+
+	// By market
+	if req.PartyId == nil && req.MarketId != nil {
+		positions, err = t.positionService.GetByMarket(ctx, *req.MarketId)
+		if err != nil {
+			return errors.Wrap(err, "getting initial positions by market")
+		}
+	}
+
+	// By party
+	if req.PartyId != nil && req.MarketId == nil {
+		positions, err = t.positionService.GetByParty(ctx, entities.PartyID(*req.PartyId))
+		if err != nil {
+			return errors.Wrap(err, "getting initial positions by party")
+		}
+	}
+
+	// All the positions
+	if req.PartyId == nil && req.MarketId == nil {
+		positions, err = t.positionService.GetAll(ctx)
+		if err != nil {
+			return errors.Wrap(err, "getting initial positions by party")
+		}
+	}
```

Comment:
Should we even allow that? Is there a use case for listening to ALL the position? isn't that a way to use loads of resources in the datanode?

Reply:
Yeah I was in two minds about this actually but I thought given the low number of markets it's not going to be orders of magnitude more data than subscribing to a whole market.. Happy to remove it though if you like, like you say, I can't really think of a use case for it outside of testing